### DISCUSSION TITLE
How can I set up a cache resolver for different inline fragments?

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#3034] [Q&A] [answered]
My `slotsToFill` query can return either a `SlotToFillForAdmin`, `SlotToFillForBuyer`, or `SlotToFillForSeller`. They all extend from the same base type, and add their own fields. I read them in using inline fragments.

<img width="609" alt="Screenshot 2023-03-13 at 11 46 27 AM" src="https://user-images.githubusercontent.com/13172299/224753977-d83cfe1d-2cbf-4103-82fa-93f4d6d30477.png">

How can I set up a cache resolver for this case? I want `id` as the key to match the existing one. These types are mutually exclusive.

Thanks!

**Comment by JoviDeCroock** [comment, marked as answer]
Checking the `__typename` is the way to go, i.e.`switch (parent.__typename)` and you can have your different cases

**Reply by nandorojo** [reply]
Thanks for the fast reply. I'm getting `parent.__typename` can only be `Query` here.

<img width="565" alt="Screenshot 2023-03-13 at 11 51 56 AM" src="https://user-images.githubusercontent.com/13172299/224755450-1002d502-0efd-4afb-ac29-b0998aaa88fd.png">

**Reply by nandorojo** [reply]
An alternative I considered: should I do something like `cache.readFragment` for each fragment and try to find the first one that exists?

**Reply by JoviDeCroock** [reply]
Ah apologies, I misunderstood the question.

```js
const isX = cache.resolve(cache.keyOfEntity({ __typename: 'x', id }));
```

Whichever rings true is probably the one that matches said id. You can probably even resolve this field from the Query type and check the __typename in-mem `cache.resolve('Query', 'slotToFill', args)`

**Reply by nandorojo** [reply]
I'm thinking the first one is better, so that it can also resolve cached items queried from a list. I'll try that. So I'll do this type of thing, right?

```tsx
const x = cache.resolve(cache.keyOfEntity({ __typename: 'x', id }));

const y = cache.resolve(cache.keyOfEntity({ __typename: 'y', id }));

const z = cache.resolve(cache.keyOfEntity({ __typename: 'z', id }));

return {
  __typename: (x || y || z)?.__typename,
  id
}
```

**Reply by nandorojo** [reply]
<img width="895" alt="Screenshot 2023-03-13 at 12 03 55 PM" src="https://user-images.githubusercontent.com/13172299/224758791-b7ad4056-170a-4bfa-bc95-083b140e941e.png">

Looks like your code example is throwing this. Is this just a type error?

Is this option below also valid?

```tsx
return cache.keyOfEntity({ __typename: 'a', id }) || cache.keyOfEntity({ __typename: 'b', id }) || cache.keyOfEntity({ __typename: 'c', id })
```

**Reply by kitten** [reply]
if you need to resolve the entity on a field: `const fieldEntity = cache.resolve(info.parentKey, info.parentFieldKey`

which is the same as: `cache.resolve('Query', info.parentFieldKey)` in your case.

Which in turn is the same as: `cache.resolve('Query', 'slotToFill', args)`

However, if you need to just find any entity by ID of a set of typenames, that's the same as doing `const entityKey = cache.keyOfEntity({ __typename, id })` where you can then check that the entity is in the cache with `!!cache.resolve(entityKey, '__typename')` since that's guaranteed to be set.

You see the pattern?
- `resolve` accesses fields
- we can use it to get data or to check that data exists
- `__typename` is guaranteed to exist if the entity is cached

**Reply by nandorojo** [reply]
Ah, got it. I misinterpreted what `keyOfEntity` did, I see that it's just creating the key without interacting with the cache now.

> if you need to just find any entity by ID of a set of typenames,

That's what I want indeed. And to clarify, `cache.resolve(entityKey, '__typename')` will return the typename itself. So I'll see if one of the type names resolve, and otherwise I'll return `null`.

```tsx
slotToFill(parent, { id }, cache, info) {
  type Slot =
    | SlotToFillForAdmin
    | SlotToFillForBuyer
    | SlotToFillForSeller
  const getTypename = <Name extends Slot['__typename']>(
    typename: Name
  ): Name => {
    return typename
  }

  const __typename =
    cache.resolve(
      cache.keyOfEntity({
        __typename: getTypename('SlotToFillForAdmin'),
        id,
      }),
      '__typename'
    ) ||
    cache.resolve(
      cache.keyOfEntity({
        __typename: getTypename('SlotToFillForBuyer'),
        id,
      }),
      '__typename'
    ) ||
    cache.resolve(
      cache.keyOfEntity({
        __typename: getTypename('SlotToFillForSeller'),
        id,
      }),
      '__typename'
    )

  if (!__typename) {
    return null
  }

  return {
    __typename: __typename as Slot['__typename'],
    id,
  }
},
```

**Reply by nandorojo** [reply]
<img width="489" alt="Screenshot 2023-03-13 at 12 28 59 PM" src="https://user-images.githubusercontent.com/13172299/224765255-0f710e75-f2c4-4109-a3ca-bd3de263b54c.png">

Awesome, that worked!

However, even though it's a cache hit with `cache-first` request policy, it still fetches every time. But I suppose that's a separate issue, so I can try to raise it elsewhere.
